sql: limit `pg_catalog` views to the current database

necaris opened the pull request:
Add `JOIN`s to the `mz_databases` view that limit `pg_catalog` views to the current database. Fixes #10391

### Testing

- [x] This PR has adequate test coverage / QA involvement has been duly considered.

### Release notes

This PR includes the following [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/developer/guide-changes.md#what-changes-require-a-release-note):

  - **TODO**:  this will probably require a release note

benesch commented:
> would appreciate some help figuring out the appropriate route for testing this. I imagine this should go in an sqllogictest but I'm not sure where yet.

We've got some existing tests for the `pg_catalog` views in https://github.com/MaterializeInc/materialize/blob/main/test/sqllogictest/pg_catalog_class.slt. I wouldn't stress about testing them too extensively, but you'll probably want to do something like create views in two different databases, and then verify that `pg_views` only tells you about the views in the current database (as controlled by `SET`/`SHOW database`).

mjibson commented on src/coord/src/catalog/builtin.rs:
```diff
@@ -1521,7 +1521,8 @@ oid,
 NULL::pg_catalog.text[] AS nspacl
-FROM mz_catalog.mz_schemas",
+FROM mz_catalog.mz_schemas
+JOIN mz_databases d ON d.id = input.database_id WHERE (d.id IS NULL OR d.name AND current_database())",
```
I don't happen to know the binding precedence for OR and AND. Light preference for some additional parens to make that explicit. Now that I look at it I'm even more confused. Is the AND anding two strings? Is that a thing?

mjibson commented on test/sqllogictest/pg_catalog_class.slt:
```diff
@@ -32,3 +32,17 @@ FROM pg_catalog.pg_class
 0 0 0 0 false p i 0 false false false false d false false
+
+statement ok
+SET DATABASE alpha;
```
`SET DATABASE TO alpha`

mjibson commented on test/sqllogictest/pg_catalog_class.slt:
```diff
@@ -32,3 +32,17 @@ FROM pg_catalog.pg_class
+
+statement ok
+SET DATABASE alpha;
+CREATE TABLE a (b int);
```
This will fail because database alpha doesn't exist.

mjibson commented on test/sqllogictest/pg_catalog_class.slt:
```diff
@@ -32,3 +32,17 @@ FROM pg_catalog.pg_class
+
+# Normal table
+query TIIIIBTTIBBBBTBB
+SELECT relname, reloftype, relam, reltablespace, reltoastrelid, relhasindex, relpersistence, relkind, relchecks,
```
I'm not clear on the goal if this query. It has `relname = 'a'` but then the result has relname as `q`.

necaris replied:
It's not clear to me how to express an _empty_ expected result, so my goal with this is to get a failure that tells me how to do it ;-)

benesch replied:
You know how to run these locally, yeah? https://github.com/MaterializeInc/materialize/blob/main/doc/developer/sqllogictest.md

You'll almost certainly want to try out the `--rewrite-results` option too!

necaris replied:
Thanks!

necaris commented:
@benesch @mjibson thanks for the comments and help so far! I've updated this and rebased into one commit for the functionality and one for an SQL logic test, which passes for me locally. Should be ready for review now!

necaris commented:
Hmm, OK, looks like the change in output columns broke the JDBC tests. I will work on those and a release note next!

benesch commented:
Ooh, you might be in for some gnarly JDBC debugging. Don't burn too long on it—I'm happy to take a look if you get stuck!

necaris commented:
Having done some JDBC debugging, it looks like no database is currently set (`SELECT current_database()` on the connection returns the empty string) so nothing comes back when filtered by it. I can see two solutions to this:
1. alter the test to explicitly set a database
2. alter the filtering logic to _not_ filter if there's no `current_database()` set, e.g. add an `OR current_database() = ''`

I have personally rarely connected to a Postgres server without specifying a database, so I would suggest that for consistency with Postgres we should go with (1). But I don't know what the conventions we're advocating around Materialize are, on this?

mjibson commented:
I'm pretty sure that in postgres you must connect to a valid database, and cannot change databases (you must reconnect if you want to do that), and queries cannot cross databases. We send a notice message if you don't have one selected, which is maybe getting ignored by jdbc here. I believe the fix is to add `materialize` after the final `/` in the connection string at https://github.com/MaterializeInc/materialize/blob/65d0561a9efa39cf1a5fbb9c1af878bdaffa3c96/test/lang/java/smoketest/SmokeTest.java#L37 which will set `materialize` as the active database.

benesch commented:
I agree! I've been wondering if we ought to do the breaking change of making it cross-database queries impossible, to leave open the possibility of having a 1:1 mapping between databases and timelines. Not sure!

necaris commented:
Great, thank you @mjibson -- I went ahead and made that change and :crossed_fingers: all the tests should pass now -- the JDBC test, at least, did locally with `mzcompose` 😄

necaris commented:
bors merge

benesch commented:
No bors over here! :D
